### ISSUE TITLE
`run0` is almost unusable.

### ISSUE THREAD

**Issue by fxzxmicah** [fedora-selinux/selinux-policy#2500]
I only tried the `restorecon` and `semodule` commands.

Log:
```
audit[108591]: AVC avc:  denied  { entrypoint } for  pid=108591 comm="(storecon)" path="/usr/sbin/setfiles" dev="nvme0n1p3" ino=754816 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:setfiles_exec_t:s0 tclass=file permissive=0
audit[116566]: AVC avc:  denied  { entrypoint } for  pid=116566 comm="(semodule)" path="/usr/sbin/semodule" dev="nvme0n1p3" ino=754814 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:semanage_exec_t:s0 tclass=file permissive=0
```

**Comment by fxzxmicah** [comment]
```
audit[29322]: AVC avc:  denied  { entrypoint } for  pid=29322 comm="(userdel)" path="/usr/sbin/userdel" dev="nvme0n1p3" ino=588622 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:useradd_exec_t:s0 tclass=file permissive=0
audit[29611]: AVC avc:  denied  { entrypoint } for  pid=29611 comm="(groupdel)" path="/usr/sbin/groupdel" dev="nvme0n1p3" ino=588611 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:groupadd_exec_t:s0 tclass=file permissive=0
```

**Comment by fxzxmicah** [comment]
```
audit[15468]: AVC avc:  denied  { entrypoint } for  pid=15468 comm="(dnf)" path="/usr/bin/dnf5" dev="nvme0n1p3" ino=924188 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:rpm_exec_t:s0 tclass=file permissive=0
audit[15549]: AVC avc:  denied  { entrypoint } for  pid=15549 comm="(dnf5)" path="/usr/bin/dnf5" dev="nvme0n1p3" ino=924188 scontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tcontext=system_u:object_r:rpm_exec_t:s0 tclass=file permissive=0
```

**Comment by septatrix** [comment]
Not sure who is at fault here but there is a similar upstream issue: https://github.com/systemd/systemd/issues/36266

**Comment by Sildulir** [comment]
So I took a look at this issue as a user and the problem seems to be on the systemd side of thing, as I explained in their issue:

https://github.com/systemd/systemd/issues/36266#issuecomment-2812350826

tl;dr: `run0` call a PAM policy that call `pam_selinux.so`, overriding the context transition that should happen with another one since the next `execve` call is the command provided on the command line.

This problem can be worked around by using `run0 sh -c '<command>'` instead of `run0 <command>`.

**Comment by poettering** [comment]
> So I took a look at this issue as a user and the problem seems to be on the systemd side of thing, as I explained in their issue:
> 
> [systemd/systemd#36266 (comment)](https://github.com/systemd/systemd/issues/36266#issuecomment-2812350826)
> 
> tl;dr: `run0` call a PAM policy that call `pam_selinux.so`, overriding the context transition that should happen with another one since the next `execve` call is the command provided on the command line.

Yes, run0 does not transition through a shell for the target binary, but invokes it directly as a sandboxed service. The policy really should be updated to deal with that. Inserting an unnecessary process in the middle, such as the suggested sh is a security weakness, and makes things less secure, not more secure.

**Comment by fxzxmicah** [comment]
@poettering I suspect there's not much to be done here, given that `run0` returns `scontext=unconfined_u:unconfined_r:unconfined_t`.

**Comment by poettering** [comment]
> [@poettering](https://github.com/poettering) I suspect there's not much to be done here, given that `run0` returns `scontext=unconfined_u:unconfined_r:unconfined_t`.

the program run0 invokes gets forked off PID 1, like a regular service. Hence, dnf should really be set up by selinux so that it can be forked off PID 1, under whatever label it lives.

**Comment by septatrix** [comment]
> the program run0 invokes gets forked off PID 1, like a regular service. Hence, dnf should really be set up by selinux so that it can be forked off PID 1, under whatever label it lives.

What I find curious is that `systemd-run --wait --collect --service-type=exec dnf --version` works whereas `run0 dnf --version` does not, so it's not completely "like a regular service"

**Comment by fxzxmicah** [comment]
```bash
#!/usr/bin/sh
RUN0="run0 --background="
if [ $# -eq 0 ]; then
	exec ${RUN0}
else
	exec ${RUN0} sh -c "$*"
fi
```
In any case, provide a wrapper script for those who come here, to ensure basic usability of run0 until either selinux-policy or systemd fully resolves the issue.

**Comment by richiedaze** [comment]
I have also run into this bug. The question I have is why does these commands by "**run0**" get blocked by selinux if it is just a link to **systemd-run** which is allowed to run?

Both are labeled the same.
```sh
lz -l /usr/sbin/{systemd-run,run0}
lrwxrwxrwx. 1 root root system_u:object_r:bin_t:s0    11 Dec 31  1969 /usr/sbin/run0 -> systemd-run
-rwxr-xr-x. 1 root root system_u:object_r:bin_t:s0 86712 Dec 31  1969 /usr/sbin/systemd-run
```

By knowing the above, bypassing becomes simple:
```sh
alias run0=systemd-run
```
Until we can discover the issue at hand.

**Comment by fxzxmicah** [comment]
> I have also run into this bug. The question I have is why does these commands by "**run0**" get blocked by selinux if it is just a link to **systemd-run** which is allowed to run?
> 
> Both are labeled the same.
> 
> lz -l /usr/sbin/{systemd-run,run0}
> lrwxrwxrwx. 1 root root system_u:object_r:bin_t:s0    11 Dec 31  1969 /usr/sbin/run0 -> systemd-run
> -rwxr-xr-x. 1 root root system_u:object_r:bin_t:s0 86712 Dec 31  1969 /usr/sbin/systemd-run
> By knowing the above, bypassing becomes simple:
> 
> alias run0=systemd-run
> Until we can discover the issue at hand.

When running the program with two different command names, `systemd-run` and `run0`, the internal execution logic of the program is different. So I don't recommend this as a workaround.

**Comment by poettering** [comment]
> > 
> 
> When running the program with two different command names, `systemd-run` and `run0`, the internal execution logic of the program is different. So I don't recommend this as a workaround.

Hm, what? no, there's a reason these two commands are implemented in the same binary. "run0" is pretty much the same as "systemd-run", it mostly just uses different defaults i.e. it spawns a service with PAM enabled by default. Hence you get the pam_selinux invocation.

We can certainly look into removing the pam_selinux invocation from systemd-run's PAM stack if that's what selinux people want.

**Comment by poettering** [comment]
> What I find curious is that `systemd-run --wait --collect --service-type=exec dnf --version` works whereas `run0 dnf --version` does not, so it's not completely "like a regular service"

run0 adds basically the implicit equivalent of `--property=PAMname=systemd-run0` to the invocation, compared to systemd-run.

**Comment by fxzxmicah** [comment]
> Hm, what? no, there's a reason these two commands are implemented in the same binary. "run0" is pretty much the same as "systemd-run", it mostly just uses different defaults i.e. it spawns a service with PAM enabled by default. Hence you get the pam_selinux invocation.

This is what I mean by "the internal logic differs when executed with different names". Maybe you have a different definition of this statement from mine, but for me, this is exactly "the internal logic differs when executed with different names".

**Comment by WavyEbuilder** [comment]
> > What I find curious is that `systemd-run --wait --collect --service-type=exec dnf --version` works whereas `run0 dnf --version` does not, so it's not completely "like a regular service"
> 
> run0 adds basically the implicit equivalent of `--property=PAMname=systemd-run0` to the invocation, compared to systemd-run.

I've looked into it and this deficiency in pam_selinux is the problem imo. Do note that on my system (not fedora policy, but still applicable), machinectl actually works perfectly and as you would expect, but the different PAM name causes a problem here. I actually think the work that needs to be done here is in pam_selinux (and as such am doing it myself) to support having different entry contexts dependent on the PAM name (in /etc/selinux/contexts/default_context and users/<various contexts apply>).

@poettering 
> We can certainly look into removing the pam_selinux invocation from systemd-run's PAM stack if that's what selinux people want.

I don't think this would help (if anything would make the situation worse). As mentioned above, I don't really think there is anything sensible that can be done from the systemd side: the best solution I see is a bit of work in pam_selinux and possibly libselinux.

**Comment by alastortenebris** [comment]
I don't know if this info is helpful or not, but on OpenSUSE Tumbleweed, `run0`, even with SELinux enabled, runs with as far as I can tell zero issues. 
I couldn't tell you what they're doing differently to make `run0` work with SELinux.

**Comment by amessina** [comment]
Maybe not the best, but I've had no issues using `run0 -i`, since bash is an allowed entrypoint.

**Comment by alastortenebris** [comment]
> Maybe not the best, but I've had no issues using `run0 -i`, since bash is an allowed entrypoint.

`--via-shell` is better as it doesn't change the working directory to your home directory.
